File: src/design_patterns/behavioral/observer/python/observer.py
class Observer:
    """ """
    def update(self, subject):
        """

        :param subject: 

        """
        pass


class Subject:
    """ """
    def __init__(self):
        self._observers = []

    def add_observer(self, observer):
        """

        :param observer: 

        """
        self._observers.append(observer)

    def remove_observer(self, observer):
        """

        :param observer: 

        """
        self._observers.remove(observer)

    def notify(self):
        """ """
        for observer in self._observers:
            observer.update(self)


class ConcreteObserver(Observer):
    """ """
    def update(self, subject):
        """

        :param subject: 

        """
        print(f"Subject's state has changed to {subject.state}")


class ConcreteSubject(Subject):
    """ """
    def __init__(self):
        super().__init__()
        self.state = None

    def set_state(self, state):
        """

        :param state: 

        """
        self.state = state
        self.notify()

File: src/design_patterns/behavioral/observer/javascript/observer.js
class Observer {
    update(subject) { }
}

class Subject {
    constructor() {
        this._observers = [];
    }

    addObserver(observer) {
        this._observers.push(observer);
    }

    removeObserver(observer) {
        const index = this._observers.indexOf(observer);
        if (index !== -1) {
            this._observers.splice(index, 1);
        }
    }

    notify() {
        this._observers.forEach((observer) => observer.update(this));
    }
}

class ConcreteObserver extends Observer {
    update(subject) {
        console.log(`Subject's state has changed to ${subject.state}`);
    }
}

class ConcreteSubject extends Subject {
    constructor() {
        super();
        this.state = null;
    }

    setState(state) {
        this.state = state;
        this.notify();
    }
}
